go/vt/vtgate/engine/distinct.go

```diff
@@ -28,37 +28,46 @@ var _ Primitive = (*Distinct)(nil)
 
 // Distinct Primitive is used to uniqueify results
 type Distinct struct {
-	Source Primitive
+	Source        Primitive
+	ColCollations []collations.ID
 }
 
 type row = []sqltypes.Value
 
 type probeTable struct {
-	m map[int64][]row
+	seenRows      map[evalengine.HashCode][]row
+	colCollations []collations.ID
 }
 
 func (pt *probeTable) exists(inputRow row) (bool, error) {
 	// calculate hashcode from all column values in the input row
-	code := int64(17)
-	for _, value := range inputRow {
-		hashcode, err := evalengine.NullsafeHashcode(value)
+	code := evalengine.HashCode(17)
+	for idx, value := range inputRow {
+		// We use unknown collations when we do not have collation information
+		// This is safe for types which do not require collation information like
+		// numeric types. It will fail at runtime for text types.
+		collation := collations.Unknown
+		if len(pt.colCollations) > idx {
+			collation = pt.colCollations[idx]
+		}
+		hashcode, err := evalengine.NullsafeHashcode(value, collation, value.Type())
 		if err != nil {
 			return false, err
 		}
 		code = code*31 + hashcode
```

Comment:
Not super important but we should probably explain why `evalengine.HashCode(17)` and why `code*31+hashcode`. I personally don't understand those arbitrary values